leader/undirected_ring/sync_higham_przytycka/sync_higham_przytycka.go

```diff
@@ -0,0 +1,413 @@
+		(curr.Round%2 == 1 && curr.Id < last.Id)
+	return drop, promote
+}
+
+func (simplePromoter) Promote(newEnvelope *uhpEnvelope) {
+	newEnvelope.Round += 1
+}
+
+// endregion
+
+// region Promoter::hp
+type hpPromoter struct {
+	F [FIB64_UPPER_LIM]step
+}
+
+func (hpPromoter) Test(last uhpEnvelope, curr uhpEnvelope) (drop bool, promote bool) {
+	drop = curr.Round == last.Round // equal rounds are a necessary condition to drop envelope
+	switch curr.Round % 2 {
+	case 0: // even
+		drop = drop && curr.Id < last.Id
+		promote = last.Round == curr.Round-1
+	case 1: // odd
+		promote = drop || curr.Cnt == 0
+		drop = drop && curr.Id > last.Id
+	}
+	promote = promote && !drop // promote implies !drop
+	return drop, promote
+}
+
+func (p *hpPromoter) fib(n int) step {
+	if n <= 1 {
+		return step(n)
+	}
+	if p.F[n] != 0 {
+		return p.F[n]
+	}
+	return p.fib(n-1) + p.fib(n-2)
+}
+
+func (p *hpPromoter) Promote(newEnvelope *uhpEnvelope) {
+	newEnvelope.Round += 1
+	newEnvelope.Cnt = p.fib(int(newEnvelope.Round + 2))
+}
+
+// endregion
+// endregion
+
+type uhpProcessor struct {
+	Id       ID
+	Dir      direction // for observability
+	Last     uhpEnvelope
+	Status   status
+	Promoter uhpPromotionStrategy
+}
+
+// endregion
+
+// region uhp:: (scope: bidirectional)
+
+// region uhp::state
+func (uhp) getState(node lib.Node) (state uhpState) {
+	json.Unmarshal(node.GetState(), &state)
+	// couldn't figure out how to serialize interface nicely
+	state.Processors[Left].Promoter = &hpPromoter{}
+	state.Processors[Right].Promoter = state.Processors[Left].Promoter
+	return state
+}
+
+func (uhp) updateState(node lib.Node, state *uhpState) {
+	payload, _ := json.Marshal(state)
+	node.SetState(payload)
+}
+
+// endregion
+
+func (uhp) send(d direction, node lib.Node, state *uhpState, msg *uhpMessage) {
+	if msg == nil {
+		return
+	}
+	if (*msg == uhpMessage{}) {
+		node.SendMessage(d, nil)
+	} else {
+		payload, _ := json.Marshal(msg)
+		node.SendMessage(d, payload)
+	}
+}
+
+func (uhp) receive(d direction, node lib.Node, state *uhpState) *uhpMessage {
+	var msg *uhpMessage
+	json.Unmarshal(node.ReceiveMessage(d), &msg)
+	if msg == nil {
+		return &uhpMessage{}
+	}
+	return msg
+}
+
+// region uhp::constructors
+func (uhp) newInstance(node lib.Node) *uhpState {
+	id := node.GetIndex()
+	res := &uhpState{
+		IsLeader: false,
+		Processors: [2]uhpProcessor{
+			uhp{}.newProcessor(id, Left),
+			uhp{}.newProcessor(id, Right),
+		},
+	}
+	return res
+}
+
+func (uhp) newEnvelope(id int, d direction) uhpEnvelope {
+	// id = id % DEBUG_ID_MOD
+	// log.Printf(LOG_PREFIX+"created envelope", id, d)
+	return uhpEnvelope{
+		Id:    id,
+		Dir:   d,
+		Round: 0,
+		Max:   id,
+	}
+}
+
+func (uhp) newProcessor(id int, d direction) uhpProcessor {
+	log.Printf(LOG_PREFIX+"created processor", id, d)
+	return uhpProcessor{
+		Id:       id,
+		Dir:      d,
+		Status:   Active,
+		Promoter: &hpPromoter{},
+	}
+}
+
+// endregion
+
+func (uhp) initiateElection(node lib.Node) {
```

Comment:
A general question: why is everything a method of uhp, instead of passing it as a param whenever necessary? Please refactor to be in line with other algorithms.

This is especially puzzling because some functions (e.g. newEnvelope, newProcessor) don't need uhp handle at all.

Reply:
There is no elegant way to namespace in Go, and this is one of the workarounds I came across, which is used even in the standard lib. I didn't want to associate the state struct with any side effects but rather as a simple dataclass, so it made more sense to pass it to 'namespaced functions' instead of calling methods as if it were a stateful object